runtime/compilers/rillv1/parse_alert.go

```diff
@@ -116,65 +117,79 @@ func (p *Parser) parseAlert(node *Node) error {
-		return fmt.Errorf(`invalid value %q for property "query.name"`, tmp.Query.Name)
-	}
+	// Data and query
+	var resolver string
+	var resolverProps *structpb.Struct
+	var queryForUserID, queryForUserEmail string
+	var queryForAttributes *structpb.Struct
+	isLegacyQuery := tmp.Data == nil
 
-	// Query args
-	if tmp.Query.ArgsJSON != "" {
-		// Validate JSON
-		if !json.Valid([]byte(tmp.Query.ArgsJSON)) {
-			return errors.New(`failed to parse "query.args_json" as JSON`)
+	if !isLegacyQuery {
+		var refs []ResourceName
+		resolver, resolverProps, refs, err = p.parseDataYAML(tmp.Data)
+		if err != nil {
+			return fmt.Errorf(`failed to parse "data": %w`, err)
 		}
+		node.Refs = append(node.Refs, refs...)
 	} else {
-		// Fall back to query.args if query.args_json is not set
-		data, err := json.Marshal(tmp.Query.Args)
-		if err != nil {
-			return fmt.Errorf(`failed to serialize "query.args" to JSON: %w`, err)
+		// Query name
+		if tmp.Query.Name == "" {
+			return fmt.Errorf(`invalid value %q for property "query.name"`, tmp.Query.Name)
 		}
-		tmp.Query.ArgsJSON = string(data)
-	}
-	if tmp.Query.ArgsJSON == "" {
-		return errors.New(`missing query args (must set either "query.args" or "query.args_json")`)
-	}
 
-	// Query for: validate only one of user_id, user_email, or attributes is set
-	n := 0
-	var queryForUserID, queryForUserEmail string
-	var queryForAttributes *structpb.Struct
-	if tmp.Query.For.UserID != "" {
-		n++
-		queryForUserID = tmp.Query.For.UserID
-	}
-	if tmp.Query.For.UserEmail != "" {
-		n++
-		_, err := mail.ParseAddress(tmp.Query.For.UserEmail)
-		if err != nil {
-			return fmt.Errorf(`invalid value %q for property "query.for.user_email"`, tmp.Query.For.UserEmail)
+		// Query args
+		if tmp.Query.ArgsJSON != "" {
+			// Validate JSON
+			if !json.Valid([]byte(tmp.Query.ArgsJSON)) {
+				return errors.New(`failed to parse "query.args_json" as JSON`)
+			}
+		} else {
+			// Fall back to query.args if query.args_json is not set
+			data, err := json.Marshal(tmp.Query.Args)
+			if err != nil {
+				return fmt.Errorf(`failed to serialize "query.args" to JSON: %w`, err)
+			}
+			tmp.Query.ArgsJSON = string(data)
 		}
-		queryForUserEmail = tmp.Query.For.UserEmail
-	}
-	if len(tmp.Query.For.Attributes) > 0 {
-		n++
-		queryForAttributes, err = structpb.NewStruct(tmp.Query.For.Attributes)
-		if err != nil {
-			return fmt.Errorf(`failed to serialize property "query.for.attributes": %w`, err)
+		if tmp.Query.ArgsJSON == "" {
+			return errors.New(`missing query args (must set either "query.args" or "query.args_json")`)
 		}
-	}
-	if n > 1 {
-		return fmt.Errorf(`only one of "query.for.user_id", "query.for.user_email", or "query.for.attributes" may be set`)
-	}
 
-	if len(tmp.Email.Recipients) > 0 && len(tmp.Notify.Email.Recipients) > 0 {
-		return errors.New(`cannot set both "email.recipients" and "notify.email.recipients"`)
+		// Query for: validate only one of user_id, user_email, or attributes is set
```

Comment:
We need to also support `Query.For` for resolvers. The resolved attributes can be passed in `ResolveOptions.UserAttributes` (needed to enforce security policies).

Maybe we should just support `For` at the top-level of the yaml? Like:
```
type: alert
data:
  ...
for:
  ...
notify:
  ...